mono/metadata/threads.c

```diff
@@ -1823,28 +1823,32 @@ ves_icall_System_Threading_Thread_Join_internal(MonoThread *this_obj, int ms)
 
 	UNLOCK_THREAD (thread);
 
-	if(ms== -1) {
-		ms=MONO_INFINITE_WAIT;
-	}
+	if (ms == -1)
+		ms = MONO_INFINITE_WAIT;
 	THREAD_DEBUG (g_message ("%s: joining thread handle %p, %d ms", __func__, handle, ms));
 	
 	mono_thread_set_state (cur_thread, ThreadState_WaitSleepJoin);
 
-	ret=mono_join_uninterrupted (handle, ms, &error);
+	ret = mono_join_uninterrupted (handle, ms, &error);
 
 	mono_thread_clr_state (cur_thread, ThreadState_WaitSleepJoin);
 
 	mono_error_set_pending_exception (&error);
 
-	if(ret==MONO_THREAD_INFO_WAIT_RET_SUCCESS_0) {
+	if (ret == MONO_THREAD_INFO_WAIT_RET_SUCCESS_0) {
 		THREAD_DEBUG (g_message ("%s: join successful", __func__));
 
-		return(TRUE);
+		/* Wait for the thread to really exit */
+		mono_threads_join_lock ();
```

Comment:
Other uses are wrapped in `MONO_ENTER_GC_SAFE`/`MONO_EXIT_GC_SAFE`. Do we need that here a well?

Reply:
Will add it.